[WIP][FLINK-23342][state] State ownership: compute shared state on recovery

rkhachatryan opened the pull request:
## What is the purpose of the change

For the changelog state backend, state ownership model
is changed from pure-JM to a mixed one, in which JM only owns
the shared state (resulting from rescaling or sharing files).

This PR adds support for such shared state management (mostly JM side):
- introduce StateObjectID (and refactor existing related classes) 1456251ea104bfa31f895928a1809dcd5117a664
- compute shared state on recovery (in `StateAssignmentOperation`) 859524ef665256a4b91bb890970098a9c26d0ac0
- inject those shared states into `TaskStateSnapshots` (TM) and `SharedStateRegistry` (JM)

To achieve this it introduces a visitor pattern to traverse StateObject hierarchies (539928c9e096c9d991d069721e6853a2cd236311).

cc: @StephanEwen, @tillrohrmann 

## Brief change log

## Verifying this change
- `SharedStateCollectorTest`
- `StateObjectIdCollectingVisitorTest`
- TBD: tests for `StateObject.accept()` implementations

This change is a trivial rework without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/JobManagerTaskRestore.java:
```diff
@@ -58,4 +78,12 @@ public String toString() {
+    }
+
+    void addSharedObjectStateID(StateObjectID sharedStateID) {
+        sharedStateObjectIDs.add(sharedStateID);
```
This set is mutable, because it's only possible to compute shared state once it is assigned.

It probably can be avoided by doing an additional pass in `StateAssignementOperation` with some "fake" graph objects or collecting the assingments, but this doesn't seem to worth it.

Zakelly reviewed (commented):
@rkhachatryan Thanks for the PR! I left some comments, PTAL. thanks.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotResult.java:
```diff
@@ -116,6 +116,14 @@ public long getStateSize() {
+    public <E extends Exception> void accept(StateObjectVisitor<E> visitor) throws E {
+        if (jobManagerOwnedSnapshot != null) {
+            jobManagerOwnedSnapshot.accept(visitor);
+        }
```
I think you are right, will add `taskLocalSnapshot` too.

(I wasn't sure because `taskLocalSnapshot` shouldn't take part for example in size calculation. However, this should be a concern of a specific visitor and it would be more clear to follow the contract of `accept()` here of visiting all children)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/RetrievableStreamStateHandle.java:
```diff
@@ -85,4 +90,10 @@ public long getStateSize() {
+    @Override
+    public <E extends Exception> void accept(StateObjectVisitor<E> visitor) throws E {
+        wrappedStreamStateHandle.accept(visitor);
+        visitor.visit(this);
```
Good point. I removed visiting the wrapped state handle to avoid double visits.

rkhachatryan commented:
Thanks for the review @Zakelly! I've addressed your feedback (dc13f075dd7bedbe99051bb7940558b1dcfbedb4, 03c811435370ac84a1374f22ccc13ab4a5715c2b)

I also updated the PR after some offline discussions with @StephanEwen (956bfbfb9912ae12bdb912a032e421fa70ffd6a6, 174563b9522340fd4ad4b2cc946bbec6daf315ec).

Zakelly reviewed (commented):
Thanks for the update @rkhachatryan , I like the idea of ```ShareableStateHandle```. Only 1 minor comment left, overall LGTM.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateObjectIdCollectingVisitor.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void visit(StateObject stateObject) throws RuntimeException {
+        if (stateObject instanceof ShareableStateHandle) {
```
You are right, I'll rename the class.
